### PR TITLE
Updated Homework #8 for 2021

### PR DESCRIPTION
**Pull request by borjasotomayor**
Homework #8 is, in substance, the same as last year, but moving to GitHub Actions has required making some non-trivial updates to the writeup:

- The homework now starts by explaining how to run a Redis server as part of a workflow, using the `services` option. This abstracts away all the details of using Docker so, at this point, they don't need to know any Docker commands.
- This is followed by the (ungraded) CS VM task, which used to happen earlier in the homework. This is where they see the commands to run Docker. We then modify the workflow file to use a custom Dockerfile.
- The Dockerfile example has changed from "running Redis with a custom module" to "building a Redis container image from scratch" (mostly because the module code does not seem to work anymore with the newer versions of Redis).
- Part II of the homework has remained largely unchanged, except for a few minor changes.

### CODE REVIEW

File: website/hw/hw8.rst [hunk 1]
```diff
@@ -638,8 +768,8 @@ Now, from your pipeline’s page, do the following:
    of the page.
```
Comment (MaxineK36) [inline website/hw/hw8.rst:768]: Very minor, but the page is now a "Review Apps" section where the actual button is "Enable"

File: website/hw/hw8.rst [hunk 2]
```diff
@@ -638,8 +768,8 @@ Now, from your pipeline’s page, do the following:
...
 -  Enable “Create new review apps for new pull requests automatically”
 -  Do NOT enable “Wait for CI to pass” (this refers to Heroku’s own CI,
-   and we are already using Travis for CI)
--  Enable “Destroy stale review apps automatically”
+   and we are already using GitHub Actions for CI)
+-  Enable “Destroy stale review apps automatically”.
```
Comment (MaxineK36) [inline website/hw/hw8.rst:772]: It lets you specify a number of days
(It says "After 5 days without any deploys", where "After 5 days" is a dropdown menu)

File: website/hw/hw8.rst [hunk 3]
```diff
@@ -638,8 +768,8 @@ Now, from your pipeline’s page, do the following:
...
 -  Don’t worry about the “Review apps may incur dyno and add-on
```
Comment (MaxineK36) [inline website/hw/hw8.rst:773]: This doesn't appear  anywhere

File: website/hw/hw8.rst [hunk 4]
```diff
@@ -651,15 +781,14 @@ Now, from your pipeline’s page, do the following:
 Now, you’ll create a pull request:
```
Comment (MaxineK36) [inline website/hw/hw8.rst:781]: The "Enable Review Apps" button appeared on the side before the checkboxes appeared on the main page

**Review by MaxineK36 — CHANGES_REQUESTED** [review]
A couple of minor clarifications, otherwise looks great!

One other note: we may want to let students know that Homework 8 depends on Homework 7; some may be planning on using an extension and would probably appreciate the heads up

File: website/hw/hw8.rst [hunk 5]
```diff
@@ -49,48 +52,284 @@ Redis server ourselves. So, if we wanted to run Redis during a CI build,
...
+into our repository.
+
+In general, the problem we’re facing is that there is some code produced
+by a “third party” that we would like to include in our project. In Git,
+this is typically handled by using *submodules*, which effectively allow
+us to include “third party repositories” in our own repository (in other
+version control systems, this is sometimes referred to as “vendor
+branches”).
+
+We can add submodules by using the ``git submodule`` command:
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst]: Done.

File: website/hw/hw8.rst [hunk 6]
```diff
@@ -49,48 +52,284 @@ Redis server ourselves. So, if we wanted to run Redis during a CI build,
...
+
+However, if we try to run the sample program, it will fail because it can't
+find a Redis server to connect to::
+
+    $ sample/sample-redis
+    Connection error: Connection refused
+
+Before we see how to launch a Redis server as part of a CI job in the next task,
+make sure you’ve successfully pushed the submodule to your GitHub
+repository. You do not need to enter anything on Gradescope.
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst:177]: Done.

File: website/hw/hw8.rst [hunk 7]
```diff
@@ -49,48 +52,284 @@ Redis server ourselves. So, if we wanted to run Redis during a CI build,
...
+see there is a new "Initialize containers" step, where GitHub Actions
+launches a Docker container with a Redis server. The "Run sample program"
+step should be able to connect to the Redis server, and you'll see this
+output in that step:
+
+.. code:: shell
+
+   PING: PONG
+   HSET: (null)
+   HGET: Alan T. Foobar
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst]: This was leftover from the "custom module" version of the homework. It should be "Random J. Redisuser" everywhere (I've updated the output in the writeup accordingly)

File: website/hw/hw8.rst [hunk 8]
```diff
@@ -49,48 +52,284 @@ Redis server ourselves. So, if we wanted to run Redis during a CI build,
...
+not have on the CS machines. So, we will use the `UChicago CS Virtual
 Machine <https://howto.cs.uchicago.edu/vm:index>`__, where you do have
-the ability to run commands with root privileges. In particular, we
-encourage you to use the VM in `Headless
-Mode <https://howto.cs.uchicago.edu/vm:headless>`__.
+the ability to run commands with root privileges.
 
 Completing this task requires running the VM on your own machine (not on
-a CSIL machine), so this task is optional and ungraded, but strongly
+a CS machine), so this task is optional and ungraded, but strongly
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst]: I've removed it to minimize confusion.

File: website/hw/hw8.rst [hunk 9]
```diff
@@ -121,13 +360,12 @@ documentation <https://docs.docker.com/install/linux/docker-ce/ubuntu/>`__):
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst:365]: Done.

File: website/hw/hw8.rst [hunk 10]
```diff
@@ -201,281 +442,170 @@ On a separate terminal, you can also try running a few Docker commands:
...
-
-Before we continue, we need to commit these submodules. If you run
-``git status``, you’ll see that Git is aware that we’ve added two
-submodules (they are tracked in a special file called ``.gitmodules``)
-but they have not yet been committed:
+So far, we've been using pre-existing container images found on `Docker
+Hub <https://hub.docker.com/_/redis/>`__, but we also have the ability
+to define our own container images. Your homework repository includes
+a ``Dockerfile`` file that includes the specification of a custom Redis
+image. To give it a try, start by cloning your repository inside the VM:
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst]: I've clarified that they should go back to their home directory if they're still inside `redis-stable`

File: website/hw/hw8.rst [hunk 11]
```diff
@@ -201,281 +442,170 @@ On a separate terminal, you can also try running a few Docker commands:
...
+    $ ./sample/sample-redis
+    PING: PONG
+    HSET: (null)
+    HGET: Random J. Redisuser
 
-{{% /note %}}
 
-Add a ``.travis.yml`` file to your repository with the following
-contents:
+Task 4: Using a Custom Dockerfile in our Workflow
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst:548]: I've added a note at the top of the task clarifying that they don't need to work on this task in the CS VM.

File: website/hw/hw8.rst [hunk 12]
```diff
@@ -558,27 +688,27 @@ your app on Heroku (just go to http://CNETID-cs220-prod.herokuapp.com/),
...
 [20 points]
 
 First, let’s make a change to our HelloApp: it’s time to upgrade to
 HelloApp 2.0!
 
--  In your repository from Lab 7 (``2020-lab7-GITHUB_USERNAME``), edit
-   ``hello/templates/base.html`` and change “Helloapp” to “Helloapp
-   2.0”.
+-  In your repository from Homework #7 (``2021-hw7-GITHUB_USERNAME``), edit
+   ``hello/templates/base.html`` and change ``<h1>HelloApp</h1>`` to ``<h1>HelloApp 2.0</h1>``
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw8.rst:714]: No, we just need them to make one change, and this ensures that the change they make is easily checkable when opening their app (if we asked them to do both, some students may update the title but then forget to update the h1)

File: website/hw/hw8.rst (hunk 1 above)
Comment (borjasotomayor) [reply website/hw/hw8.rst:779]: Is the button actually "Enable"? It shows up as "Enable Review Apps" on my end.

Screenshot from 2021-05-15 10-32-02

File: website/hw/hw8.rst (hunk 2 above)
Comment (borjasotomayor) [reply website/hw/hw8.rst]: Clarified that they can leave the default value selected.

File: website/hw/hw8.rst (hunk 3 above)
Comment (borjasotomayor) [reply website/hw/hw8.rst]: It's easy to miss, but it appears above the "Enable Review Apps" button (I've updated the text to clarify; the previous text sounded like they would see a pop-up or something like that)

Screenshot from 2021-05-15 10-33-21

File: website/hw/hw8.rst (hunk 4 above)
Comment (borjasotomayor) [reply website/hw/hw8.rst:795]: I'm not sure what this refers to. What checkboxes do you mean?

File: website/hw/hw8.rst (hunk 1 above)
Comment (MaxineK36) [reply website/hw/hw8.rst:779]: @borjasotomayor Ah, I see. I was on the settings page. From the pipeline homepage, I see exactly what's in the image above.

File: website/hw/hw8.rst (hunk 4 above)
Comment (MaxineK36) [reply website/hw/hw8.rst:795]: Nevermind, this was due to my being on the settings page.

**Review by MaxineK36 — APPROVED** [review]
Looks great!

**Review by ecoble — APPROVED** [review]
LGTM! I'm not sure if this has already been explained in class, but I have found that a lot of students don't know what an API is, so it might be good to have a brief explanation when it's first brought up or at least provide what the acronym stands for, but also I don't think it's necessary for them to complete the homework.

**Comment by borjasotomayor** [comment]
The term "API" is introduced in the lectures, but kind of in passing. I've made a note to talk about APIs in a bit more detail next year, but I've also expanded the API acronym the first time it appears, so that students can connect it with the notion of an interface.

**Pull request merged**